### DISCUSSION TITLE
FNIRSI GC-01: Can't flash firmware on "AIR01" device

### DISCUSSION THREAD

**Discussion by nefarium** [Gissio/radpro#45] [Q&A] [answered]
I am able to connect via USB after careful timing of holding and releasing both right arrow and power button.  However, the counter only shows up as "AIR01" under windows 10, Devices and Printers (unspecified subgroup).  Cannot access it's files and it's not showing up as a drive or anything useable.  Any suggestions?  Thank you.

**Comment by Gissio** [comment]
Strange.

Can you post a screenshot of the device in "Device Manager"? In the device properties, can you go to "Details" and post what you see under "Hardware Ids" and "Capabilities" (or "Capacities")?

Can you try on another Windows computer?

**Reply by nefarium** [reply]
Yeah, if I remember right it was just showing up as an unknown device with no helpful data under device manager details.  I will get some screenshots taken later and forward them.  I tried it on total of three computers.  The other ones are an older Win7 laptop and also a Win Vista (cringe) laptop.  My main rig is Win 10 which should not have any issues.  One tidbit of information I noticed is that when inspecting the WCH microchip under magnifying lens, I could see that the final two digits of the chip number were "10" instead of "20" like shown in some online pictures.  So, perhaps an earlier version of the chip.  Thanks.

**Reply by nefarium** [reply]
OK I can see under device manager description/general tab the following text for the USB Mass Storage Device which appears on GC power up: "This device is working properly. There is a secondary device connected to this hardware that Windows cannot identify because it does not have a valid hardware identification number. For assistance, contact the hardware manufacturer."  So.... I spent some time and combined several screenshots you requested into one MS paint image.  You can see the "AIR01" which appears when the GC is powered up bottom left with big red arrow pointing to it.  Smaller red arrow near center points to it in the Device Mananger.  The two smaller boxes on right side of screenshot are the Hardware IDs and Capabilities shots.  Let me know if you see anything helpful there or not.  
![screenshots](https://github.com/Gissio/radpro/assets/164772133/05d9fb4a-ab25-4280-9dae-019b20ed0c5d)

**Reply by nefarium** [reply]
PS: I also took a screenshot of the details provided by an application called USBDeview, if this data helps you at all.  Thanks.  
![USBDeview](https://github.com/Gissio/radpro/assets/164772133/d3a0291c-33d0-45cc-9cc8-6569d673d749)

**Comment by Gissio** [comment]
So what happens if you drag the Rad Pro firmware to the "AIR01" icon? Make sure you [check what MCU](https://github.com/Gissio/radpro/blob/main/docs/devices/FNIRSI%20GC-01/install.md#step-1-open-the-device) your GC-01 has, and upload the corresponding firmware.

**Reply by nefarium** [reply]
Cannot really tell which version of the firmware because the power screen only shows the FNIRSI logo / website link and some chinese letters.  No firmware version shown there.  I made a video of it and paused it to make sure I didn't miss the version.  It's just not there.  The chip is the WCH chip and as noted the last two digits of it's number are 10 instead of 20 like I see online.  I can take it apart again if you think it helps to provide the full long chip number.  Mine came with the M4011 geiger tube which I don't see mentioned very much except as an option.  The M4011 tube seems like a decent one used by GQ electronics, etc.  Thanks.

**Reply by Gissio** [reply]
If you have a WCH MCU, you can recover with the original V1.5 firmware. If you have a Geehy MCU, you can recover with the original V1.6-2 firmware.

Did you try dragging the GC-01 Rad Pro firmware for WCH to the "AIR01" icon?

**Reply by nefarium** [reply]
OK Cool.  I did not want to try that in fear of bricking the device.  However, it's not a huge loss if it happens because I only paid about $45 delivered for it.  I am heading off to work now but can try it this evening.  I have a feeling it's not going to work because it does not show up as a valid storage device like a USB thumb drive does.  Will give it a shot though.  Thanks again.

**Reply by Gissio** [reply]
Some storage devices do not show up as regular drives (iPhones, for example). Let us know how it worked.

**Reply by Lion22332** [reply]
Same here: the chipnumber is CH32F104 R8T6 401546C10. And the tube a m4011.  I bought it may 2023. Windows recognises the Fnirsi with a dongle sound when you push the middle button. If you let the button go then there is a  the disconnect sound. Under device manager there is a mass storage device listed under USB and it is said to be working properly . The driver is ..system/drivers/usbccgp.sys. And also an AIR01 is listed in bluetooth and other devices in the settings. Somewhere in the settings there was aan option to create a storage space on it and is warned that the files (a text file) will be deleted when you do that. But the basic problem i think is that there is no or a wrong driver for it installed. I could not drag the Radpro firmware anywhere to this drive as it soes not show up in win explorer. :(

**Comment by Gissio** [comment]
To both of you, try this:

* Open "Disk Management".
* In the "View" menu, select "Volume" and "Disk List".
* Connect your FNIRSI GC-01 and keep the power button pressed.
* A disk should appear in the top view. Click on the disk (first column).
* Post a screenshot of what you see.

**Reply by Lion22332** [reply]
I looked there already but it seems nothing shows up. Maybe if you create a storage space on it as i have seen somewhere before and thus you loose all files you can copy the bootloader program together with  the Radpro but i don't want to try that and brick my Fnirsi LOL.

**Reply by nefarium** [reply]
OK I tried it too and no new disk shows up.  Also tried just dragging and dropping your bin file to the AIR01 icon and it does not allow it.  Shows a little crossed out circle to indicate it's not possible when I try.

**Reply by nefarium** [reply]
I wonder if contacting FNIRSI support is worth a try to ask if the WCH version with M4011 geiger tube even has the ability to receive firmware updates like the other models seem to(?)  Shot in the dark perhaps.

**Comment by Gissio** [comment]
Honestly I don't know what to do about this.

**Reply by nefarium** [reply]
Been a busy week at work but tomorrow is my day off so I can mess with it some more.  Will report back.  Thanks.

**Comment by Lion22332** [comment]
Could it be that the problem is that the processor in my GC-01 is not CH32F103R8T6 but CH32F104R8T6?

**Reply by Gissio** [reply]
According to WCH ([https://wch-ic.com/products/categories/66.html?pid=5](https://wch-ic.com/products/categories/66.html?pid=5)), there should be no CH32F104R8T6. Can you post a photo of the MCU?

**Reply by Lion22332** [reply]
Sorry, i made a mistake it is 103 after all.

**Comment by Gissio** [comment, marked as answer]
I've added instructions for flashing Rad Pro to the FNIRSI GC-01 with an STLINK USB dongle: https://github.com/Gissio/radpro/blob/main/docs/devices/FNIRSI%20GC-01/install-stlink.md

**Comment by nefarium** [comment]
Great news thank you.  Will give it a shot.  I just ordered the dongle for $5.69 which gets good reviews and some pin headers so I don't have to scavenge them.  Will be a few days to arrive but I will let you know how it goes.  PS: in your directions you don't mention the M4011 geiger tube which is also on some units.  It's a good tube also used by GQ Electronics in some of their counters.  Mine is dated 2023.  Here is a picture online for comparison: https://ae01.alicdn.com/kf/S234ea8cc10b64df2b9d77192152ddc85c/J321-Geiger-Muller-tube-counter-Hard-Beta-GM-Detectors-Geiger-Counter-Kit-The-tube-for-Nuclear.jpg

Thank you!
________________________________

**Reply by Gissio** [reply]
Thanks for reporting about the M4011.

Let us know how it went! I'm particularly interested in your experience with the haptic pulses feature. Are pulses too weak/too strong?

**Comment by nefarium** [comment]
Just to let you know the dongle arrived and worked great.  The flash worked very well on my WCH, M4011 tube unit.  One thing you may want to add is for the user to give it some time for the flashing to complete because I started to think it had froze up in a loop.  After a bit it completed fine.  Another thing is the line in the instructions: "To install from Windows, start the install-fnirsi-gc01-[mcu].bat" ... briefly confused me because the filename started with "flash" instead of "install".  Not a big deal and I figured it out.  My header pins were indeed a different layout, but no biggie.

The clicker is still pretty quiet so I may end up trying the LED/resistor in parallel mod to see if that helps.  When switched to high volume it uses the beep tone which I don't really like.  I prefer good old-fashioned clicks.

I drilled a 3/16" hole in the case for the piezo sounder to hopefully let more audio out and super glued a bit of stainless pipe screen mesh inside to keep debris/lint out.  Hoped it would help click volume but still pretty quiet.

I like Dusk mode best for now!  🙂  Cool stuff.  Thanks again!

________________________________

**Reply by Gissio** [reply]
Thanks for the comments.

Instructions were fixed. Regarding the "quiet" volume, could you try this test release:

[radpro-flashtool-2.0rc4-test2.zip](https://github.com/Gissio/radpro/files/15135960/radpro-flashtool-2.0rc4-test2.zip)

It should sound stronger, but a bit more beep-ish than click-ish.

Are you able to access the bootloader now? If you are not, could you please ask the people in this thread to be generous and post their bootloader (in particular, the ones that identify as "GC01 BOOT" drives)?: #59

**Comment by nefarium** [comment]
Been pretty tired after work but I will try it soon.  Part of the reason I don't like the beep tone is that it's too long.  When you are measuring a source like vintage orange pottery that has a high enough count rate, all of the beeps flow together into a continuously annoying sputter tone.  Clicks seem better at handling moderately high count rates.  I took the unit with me into a prospecting supply store to show the owner and was measuring a bunch of rock specimens he had sitting around.  Sadly, the click volume was so low that we could barely hear it over the ambient background noise of the fluorescent lighting above and street traffic.  And this is with a 3/16" hole drilled over the piezo sounder.  So yeah, I'd like to increase it.  The beeps were a little better in volume but not much.  I just plugged it in via USB a few different times, powering it on and off.  It still makes the USB connection sound but nothing shows up as a storage device so I don't think I have bootloader access.  I don't know what you mean about asking people to be generous and post their bootloaders...(?)  At least I can still take it apart and do the firmware flash manually.  Bit of a pain to do but if you think it may help click volume it's worth a try.

-Chris
________________________________

**Reply by Gissio** [reply]
I mean the following: there are many bootloaders in the wild. I've asked people to dump their bootloaders with the instructions given in #59, but so far only one person did this, and that's the bootloader I'm packaging with the `flash-tool`.

So it would be great if other people could do this as well, as the only bootloader we've managed to dump doesn't work on all GC-01s (as yours, for example).

**Reply by Lion22332** [reply]
As a newbie i am trying to understand the flashing with a ST-dongle on the FNIRSI.  So you just have to double click (from Windows Explorer) the mentioned .bat file from the downloaded Radpro Flashtool and it will automatically install? 

Furthermore the following sentence in the:  'Step 3 Flash The Firmware" instructions i do not understand: "To flash the original firmware, open a terminal and start the script with the firmware/fnirsi-gc01-ch32f103r8t6-v1.5.bin argument if your device has a CH32F103R8T6"

What terminal do you mean? A windows Powershell window? And what do you mean with "the script"? Is that a batch file? Or is there some software needed to do this?

**Reply by Gissio** [reply]
Regarding your first question: that is correct!

Regrading the other questions: A Windows PowerShell window is a terminal. A batch file is a script. There is no other software you need.

**Reply by Lion22332** [reply]
Thank you for your answer. Sorry for my ignorance but i still don't understand what script is meant here. Is it the batchfile that installs Radpro and that i have to adapt or is there another file that starts with the "firmware/fnirsi-gc01-ch32f103r8t6-v1.5.bin argument". Or do i have to write a new file that starts with that argument? That seems quite complicated when i see the installation batch file. I realize that you cannot pre-chew everything for us but i want to   know how to reinstall the original firmware in case i brick my Fnirsi in the installation process.

**Reply by Gissio** [reply]
Sorry I wasn't clear enough.

In computing, a script is a series of simple instructions that automate a process. You can learn more about scripts here: [https://en.wikipedia.org/wiki/Scripting_language](https://en.wikipedia.org/wiki/Scripting_language).

In order to install the original firmware, select the decompressed "radpro-flashtool" folder in Explorer, right click on the folder while pressing the shift key, select "Open PowerShell window here...", and type the following command:

    flash-fnirsi-gc01-ch32f103c8t6.bat firmware/fnirsi-gc01-ch32f103r8t6-v1.5.bin

That's all.

**Reply by Lion22332** [reply]
Thanks, it is clear now, i got everything and the ST drivers are installed. I hope soon to have the courage to try and install Radpro LOL.

**Comment by zhang00963** [comment]
This issue is a USB code design problem in the bootloader. I have fixed the relevant problem in this bootloader and it can display GC01 BOOT disk devices normally and perform firmware updates, but please be careful!! ， The bootloader contains one-to-one encrypted data, and different devices need to recalculate the verification code data after flashing. This bootloader is only applicable to FNIRSI-GC01 of CH32F103R8T6 MCU. @Gissio @nefarium 
[fnirsi-gc01-ch32f103r8t6-bootloader-revise.zip](https://github.com/user-attachments/files/24330785/fnirsi-gc01-ch32f103r8t6-bootloader-revise.zip)
md5:200b8601652a6228c1e128c643676ca3

**Reply by Gissio** [reply]
~~If this bootloader is for APM32 devices,~~ If you install this bootloader on APM32 devices, make sure you **do not** overwrite 0x08003800-0x08003fff, as I recently learned FNIRSI stores the high voltage PWM parameters there.

**Reply by zhang00963** [reply]
The APM32 version MCU stores configuration information starting from address 0x08003800, while the CH32 version MCU stores configuration information starting from address 0x08003C00, both starting with a fixed 0xEB and containing internal configuration information such as screen brightness.

**Reply by Gissio** [reply]
I was just warning users.

**Reply by Gissio** [reply]
A request: Could you verify the Chinese translation is correct?

**Reply by zhang00963** [reply]
Version 3.0.3 firmware?

**Reply by Gissio** [reply]
Yes, please. You can find the strings here: https://github.com/Gissio/radpro/blob/main/platform.io/src/strings/zh_CN.h

**Reply by zhang00963** [reply]
Of course, I will upload the file after checking and correcting it. However, the GC-01 I have cannot install the Chinese firmware I compiled because the generated firmware length exceeds 0xC000

**Reply by Gissio** [reply]
Greatly appreciated.

Just to clarify: The only device that does not support CJK languages is the GC-01 (CH32). The GC-01 (APM32) has sufficient flash memory and does support them.

**Reply by zhang00963** [reply]
Yes, the CH32 version of GC01 only has 64K space. When I have time, I will replace it with an APM32 chip. The modified language file has been uploaded here,
[zh_CN.zip](https://github.com/user-attachments/files/24354127/zh_CN.zip)

The following is the definition of the original firmware configuration area for GC01：

// CH32F103R8T6  0x08003C00 ~ 0x08003C2F
EB    // Fixed Head
09    // Brightness
07    // Alarm mode bit1 Sound bit2 Vibration bit3 Indicator light
20    // Reserved
01    // Interface Language 01 English 02 Chinese
00    // Keep default entry interface 00 Home 01 Graph 03 Settings
01    // Unit setting 01 uSv/h 02 uGy/h 03 mR/h 04 CPS 05 CPM
00    // Retain the influence of interface style
ACF4CBC21FA98A2AFC64F6B1      // chip one-to-one authentication data
716C0000     // Cumulative dose
6400             // Current dose alarm value
D007            // Cumulative dose alarm value
E6070C1F     // Start date 0x07E6=2022 (year) 0x0C=12 (month) 0x1F=31 (date)
AB840100    // Accumulated time 0x000184AB=99499 (device running time in seconds)
00000000    // Reserved
2D0000       // Keep 45
200000       // Keep 32
0000           // Reserved


// APM32F103RBT6  0x08003800 ~ 0x0800382F
EB     // Fixed Head
09     // Brightness
07     // Alarm mode bit1 Sound bit2 Vibration bit3 Indicator light
00     // Reserved
01     // Interface Language 01 English 02 Chinese
00     // Keep default entry interface 00 Home 01 Graph 03 Settings
01     // Unit setting 01 uSv/h 02 uGy/h 03 mR/h 04 CPS 05 CPM
00     // Retain the influence of interface style
000000000000000000000000 // Reserved
1A010000     // Cumulative dose
6400            // Current dose alarm value
D007           // Cumulative dose alarm value
E6070101    // Start date 0x07E6=2022 (year) 0x01=01 (month) 0x01=01 (date)
C3020000    // Accumulated time 0x000002C3=707 (device running time in seconds)
409c            // PWM frequency 40.000KHz
F401            // PWM duty cycle 50.0%
FFFFFFFFFFFFFFFF // Keep

**Reply by Gissio** [reply]
zhang00963: I just noticed your answer, sorry for the delay! Thanks for the information and for your contribution.

**Reply by Gissio** [reply]
zhang00963: there are some people (#346) claiming that Rad Pro 3.0+ doesn’t install correctly on the FNIRSI GC-01/APM32. Instead of booting correctly, they say the device enters the bootloader and displays a NOAPP.txt file.

I don’t have a GC-01/APM32, but flashed the APM32 bootloader to a development board, and then installed Rad Pro 3 through USB without issue.

You seem to be acquainted with the bootloader. What do you think could be wrong?

**Reply by zhang00963** [reply]
I have replaced the CH32F103R8T6 of GC-01 in my hand with APM32F103RBT6, and the test shows that 3.0.2 can be installed normally. Use fnirsi-gc01-apm32f103rbt6-botloader.bin from the project.

**Reply by Gissio** [reply]
Cool! Thanks for the report.

On my development board I was also able to install it through USB, so maybe there is some condition in the bootloader that does not match with Rad Pro 3.

**Reply by Gissio** [reply]
zhang00963: what changes did you apply to the file [fnirsi-gc01-ch32f103r8t6-bootloader-revise.zip](https://github.com/user-attachments/files/24330785/fnirsi-gc01-ch32f103r8t6-bootloader-revise.zip) you uploaded earlier?

**Reply by zhang00963** [reply]
Fixed the issue where the original botloader cannot display GC01 BOOT disks on some computers or devices due to incorrect content in the USB device descriptor. As for the authentication data for 3C08~3C13, it needs to be recalculated based on the chip UID using a specific algorithm. This data error can cause the original firmware to display characters incorrectly.

**Reply by Gissio** [reply]
Got it! Will be added to the repository.

**Reply by zhang00963** [reply]
According to the testing of the APM32 version boot, there may also be a similar issue where the code directly uses the chip UID as the USB descriptor Serial field data. However, in the standard, the Serial field should be a readable character, and the chip UID may not necessarily be a readable character, which may cause some devices to not display boot disk issues.

Original descriptor:
<img width="805" height="728" alt="image" src="https://github.com/user-attachments/assets/ce8439c4-6ece-4061-9884-a5c1d7a5f41c" />

Corrected descriptor:
<img width="815" height="711" alt="image" src="https://github.com/user-attachments/assets/266bcae7-f9b2-412d-be21-012376e64ea5" />

**Reply by Gissio** [reply]
Great find!

But I can’t see how this could explain that for some GC-01 / APM32 users Rad Pro 2 works but not Rad Pro 3. On my development board Rad Pro 3 works flawlessly, as well as on your “APM32-transplanted” GC-01. Strange.

**Reply by Dobfek** [reply]
@Gissio, @zhang00963 if it fixes the GC-01 USB flashing, how can I flash it onto my WCH chip? Would it be possible to flash not just the application but the bootloader too? (It would be nice to have my original one as backup first.)

**Reply by Dobfek** [reply]
Okay, now I have checked the 3.1beta2, and it already uses the bootloader from @zhang00963 . And it works well on my GC-01! :)

To fix this issue on a GC-01(ch32f103r8), you should follow these steps:
Pre-requirement:
Installing new bootloader once: Install 3.0.2(or later) over ST-LINK. It will install the bootloader from @zhang00963, which fixes the USB flasing.

Installing steps after the new bootloader on Win10:
- Plug the GC-01 into a machine over USB.
- Press power button on device until "USB device plugged in" sound can be heard or device shows up as a harddrive under the name GC01 BOOT.
- Open the drive, Readme.TXT with 0 bytes will show up.
- Copy the firmware .BIN here.
- After the copy, you will hear a "USB device unplugged unexpectedly" sound.
- Unplug the USB cable from GC-01.
- Power on the device, and be happy :)

**Comment by nefarium** [comment]
Thanks guys.  I've not messed with it since my successful flash update about a year and a half ago now.  The device has worked OK since then but I only used it a handful of times.  Took it to the antique mall a couple times with friends, etc.  I am reluctant to try another flash update since the procedure seemed to "freeze up" with no progress indication long enough that I started to think it had failed.  For a few minutes I thought I had bricked the device but it succeeded. I don't want to mess it up by trying something experimental now. I never got around to installing the last update which was supposed to improve click volume... so I cannot say if it would.  I think it was 2.0(?)...  I would need to locate the USB dongle / pin header thing which is hidden somewhere around here in my office.  🙂

Best regards,
-Chris

________________________________
